### PR TITLE
Atomic builtins

### PR DESCRIPTION
**Pull request by moskvanaft**
Building WASM-code sometimes fails (llc is unable select instructions for some operations). Fix for that.

### CODE REVIEW

File: include/builtins.hpp [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+#include <type_traits>
+/*
+template<class T> T __sync_fetch_and_add(T* ptr, T val) {
+   T old = *ptr;
+   *ptr += typename std::conditional<std::is_pointer<T>::value, size_t, T>::type( val );
+   return old;
+}
+
+template<class T> T __sync_add_and_fetch(T* ptr, T val) {
+   return ptr += typename std::conditional<std::is_pointer<T>::value, size_t, T>::type( val );
```
Comment (gleehokie) [inline include/builtins.hpp]: We should use one style for these three functions.

File: include/builtins.hpp [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+inline int32_t __sync_fetch_and_add( int32_t* ptr, int32_t val ) {
+   auto old = *ptr;
+   *ptr += val;
+   return old;
+}
+
+inline int32_t __sync_add_and_fetch( int32_t* ptr, int32_t val ) {
+   return *ptr += val;
+}
```
Comment (gleehokie) [inline include/builtins.hpp:49]: Why the different style?

File: src/locale.cpp [hunk 3]
```diff
@@ -664,10 +665,12 @@ locale::id::__get()
     return __id_ - 1;
 }
 
+#include <builtins.hpp>
```
Comment (gleehokie) [inline src/locale.cpp]: Do you really want the #include this far down in the file?  I know it is legal, but it makes it hard to see what include files are used with this file.  If there is a reason for the include to be this far down then Ok.

**Comment by spoonincode** [comment]
It seems like the compiler can possibly call out to any of the functions defined in lib/Target/WebAssembly/WebAssemblyRuntimeLibcallSignatures.cpp

Are we going to keep playing whack a mole? Why are we implementing these functions here vs in a helper library that is linked to our wasms? Why not use compiler-rt? Why are we implementing some of those compiler generated functions as intrinsics instead of code like is done in this PR?

File: src/locale.cpp (hunk 3 above)
Comment (moskvanaft) [reply src/locale.cpp]: There was no real reason. I just put it there when I was debugging the code. Moved up.

File: include/builtins.hpp (hunk 2 above)
Comment (moskvanaft) [reply include/builtins.hpp:49]: @gleehokie , I am not sure that I understand what do you mean.
If your comment is about  __sync_fetch_and_add vs  __sync_fetch_and_add - they are different functions (first one returns old value while the second returns new value). libc++ needs them both in order to compile.

**Pull request merged**